Bugfix Issue 11586 - Marketing Email gets send multiple times when using translations

bergwerkweb opened the pull request:
<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 5.x)
* a.b for any bug fixes (e.g. 4.4, 5.1)
* c.x for any features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 5.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [Y ]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [N ] <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      | mautic/mautic-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #11586 <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

#### Description:

<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do. If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->

#### Steps to test this PR:


<!--
This part is really important. If you want your PR to be merged, take the time to write very clear, annotated and step by step test instructions. Do not assume any previous knowledge - testers may not be developers.
-->
Step 1: Create a Template E-Mail (Language English)
Step 2: Create a Translation for that E-Mail (e.g. Language German)
Step 3: Create a Test Segment and add Users with "preferred language"="de" AND "preferred language"="en"
Step 4: Create a "Campaign 1" which triggers sending of email from Step 1
Step 5: Create a "Campaign 2" which, after 2 minutes, also triggers sending of email from Step 1

Every E-Mail is now only sent once.
<!--
If you have any deprecations, list them here along with the new alternative.
If you have any backwards compatibility breaks, list them here.
-->

kuzmany committed:
Add unit tests

patrykgruszka committed:
Multiselect in contact field value campaign condition

patrykgruszka committed:
LeadFieldRepository functional tests

patrykgruszka committed:
Fix accessing container in LeadFieldRepositoryFunctionalTest

kuzmany committed:
Fix loop doctype if HTML contain unicode chars

patrykgruszka committed:
CS Fixer

patrykgruszka committed:
Fix PHPSTAN

SalemCode8 committed:
Working on list view from CoreBundle

allcontributors[bot] committed:
docs: add putzwasser as a contributor for bug (#11617)

* docs: update README.md [skip ci]

* docs: update .all-contributorsrc [skip ci]

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

SalemCode8 committed:
Working on form

allcontributors[bot] committed:
docs: add beetofly as a contributor for userTesting (#11619)

* docs: update README.md [skip ci]

* docs: update .all-contributorsrc [skip ci]

Co-authored-by: allcontributors[bot] <46447321+allcontributors[bot]@users.noreply.github.com>

SalemCode8 committed:
* Finish Security Section

* Finish User Section

* #TPROD-329 - Migrate UserBundle to twig templates

* fixcs + phpstan

* mauticTemplateVars are not set for Twig but its content is

So we can get mauticContent directly

* Fix Role Form View

* Fix elements alignment

* Fix Inputs alignment on mobile

Co-authored-by: John Linhart <admin@escope.cz>
Co-authored-by: Ruth Cheesley <ruth.cheesley@acquia.com>

mollux committed:
add custom PHPCS fixer to prevent useless MAUTIC_TABLE_PREFIX defines (#11622)

* add custom PHPCS fixer to prevent useless MAUTIC_TABLE_PREFIX defines

* Fix the tests

* CS fix in CS fixer

annamunk committed:
Select with Mautic tokens for GrapesJS text editor (#11525)

* Select with Mautic tokens for GrapesJS text editor

* Form tokens in GrapesJS text editor

* GrapesJS scripts rebuilt

Co-authored-by: John Linhart <admin@escope.cz>

pwned555 committed:
Add orderBy to findOneBy to return more accurate eventLog result

escopecz committed:
Removing CustomFormExtension as it is not used and was deprecated in 4.3 (#11632)

github-actions[bot] committed:
Auto-update GrapesJS generated JS dist files (#11631)

Co-authored-by: dennisameling <dennisameling@users.noreply.github.com>

fedys committed:
Integration bundle improvements (#11303)

* Validate fields in integration

* Fixed existing tests

* Unit tests

* CS fixes

* CR fixes

* Resolver bug fixed

* Fixed unit test

* Notification deduplication and bulk notifications

* Field validation messages made more clear

* Tests fixed
CS fixes

* phpDoc improved

* CR fixes

* Fixed unit test

* Validation rules made stricter

* Renamed test method

* SchemaDefinition updated

* CS fixes

* PHPStan fixes

* Store the ObjectMapping entity fetched when updating objects

* Generate an OrderResultsDAO and dispatch an event after each batch so that plugins can take additional action on that batch if/as needed

* Each direction handles ObjectMappings differently so had to account for that when building the OrderResultsDAO.

* Expose the integration name since this event will dispatch for all integration syncs and rename getOrderStatus to getOrderResults

* Fixed setting deleted object

* Handle when ObjectMappings were not found when attempting to update

* Tests for OrderResultsDAOTest

* Added tests for updating mappings

* Tests for CompletedSyncIterationEvent

* Test OrderExecutioner::execute returning ObjectMappings

* Tests for ObjectMappingsDAO

* Add tests for the new batch sync events

* Update a new copyright date and replaced $this->assert* with Assert::assert*

* Fixed a getter loop

* Add tests to prevent decreased coverage

* Increase coverage for OrderExecutioner

* Increase coverage for new lines introduced in SyncProcess

* Increase coverage for new lines introduced in OrderExecutioner

* PHPStan fixes

* PHPStan fixes

* CS fixes

* Added missing service definitions

* 'url' type specified in SchemaDefinition

* Fixed SchemaDefinitionTest

* Fixed SyncProcessTest

* SchemaDefinition string length reverted

* Fixed LeadFieldRepositoryTest

* Added DB migration

* Removed 'charLengthLimit' from the select section

* CS fix

* CS, PHPStan, Rector and test fixes

* Test fixed

Co-authored-by: Alan Hartless <alan@devkardia.com>
Co-authored-by: Ruth Cheesley <ruth.cheesley@acquia.com>

SalemCode8 committed:
Finish Form View

SalemCode8 committed:
Add Form View

SalemCode8 committed:
#TPROD-310 Migrate ChannelBundle to twig templates

SalemCode8 committed:
fixcs + phpstan

SalemCode8 committed:
Fix variable types

SalemCode8 committed:
Fix fixcs errors

SalemCode8 committed:
Fix message translation

SalemCode8 committed:
Revert "Fix message translation"

This reverts commit 352ceea410c27e15f0c51ae63b630223cb0c0dda.

SalemCode8 committed:
Check file markup

SalemCode8 committed:
Check file markup

SalemCode8 committed:
Fix View Issues

npracht committed:
Fix/change hubspot api auth (#11470)

* Add support for bearer token from private apps

* Fix API  V2 error response

* Fix If keys are load directly from appendForm in integration

* Fix PHP stan

* Remove readonly statement

* Fix PHP Notice - Trying to access array offset on value of type null

Co-authored-by: Zdeno Kuzmany <zdeno@kuzmany.biz>

biozshock committed:
Tests for #11767.

patrykgruszka committed:
remove header comment in test

patrykgruszka committed:
remove unused import in LeadFieldRepository

patrykgruszka committed:
fix multiselect contact field condition

patrykgruszka committed:
clenup formatting

patrykgruszka committed:
fix tests

escopecz committed:
Fixing marketing message pagination

escopecz committed:
Fixing channel message clone

escopecz committed:
STAN fixes

bradjones1 committed:
Clarify segment tooltip

escopecz committed:
Temporary fix for a failing campaign test

as it was using standard file rather than the PHP file and so it was missing some content

escopecz committed:
Finishing up the twig refactoring for the marketing message campaign action template

escopecz committed:
Preparing a Twig template replacement once a Twig function is available from ohter open PR

escopecz committed:
Removing unused PHP templates

escopecz committed:
Removing unused code

kuzmany committed:
Fix reply for email with suffix

escopecz committed:
Fixing notice that "$inline" is undefined

escopecz committed:
Temporary fix when a PHP templates tries to load a TWIG template so it would render the HTML and not just echo the template content

escopecz committed:
Finishing the queued_messages template refactoring in the contact timeline

escopecz committed:
WIP channel timeline template refactoring - currently fails

dadarya0 committed:
Fix multibyte char issue

ynreob committed:
bugfix Issue 11586

dadarya0 committed:
cut string based on byte

escopecz committed:
Twig syntax fixes

escopecz committed:
Copying changes from the original offline.php to the scaffold one

escopecz committed:
Returning a bad change that got in with the previous merge.

volha-pivavarchyk committed:
add a clearer message when importing contacts

volha-pivavarchyk committed:
add a flash message if textStatus is error

escopecz committed:
Fixing marketing message list page by using if closer to the PHP version

escopecz committed:
Setting twig template is not necessary. It's done automatically

escopecz committed:
Making the if same as in the PHP version

escopecz committed:
Converting API bundle PHP templates to Twig

escopecz committed:
Removing old unused PHP templates

escopecz committed:
Fixing pagination in the list of contacts on the channel detail page

escopecz committed:
Improving formating

escopecz committed:
Fixing hiding the channel select if the channel is disabled right after save

RCheesley commented:
@cla-bot check 

Thanks for the PR @bergwerkweb and for getting the Contributors Agreement signed! Welcome to the community! 🚀

RCheesley committed:
Merge branch '5.x' into issue11586

biozshock committed:
Revert b1bb8f4fe6016c1fab98

biozshock committed:
Fixes.

kuzmany committed:
Refactor with EmailAddressHelper usage

kuzmany committed:
Add array_unique

kuzmany committed:
Remove Reply class from EmailAddressHelper.php

kuzmany committed:
Add unit tests

kuzmany committed:
Add mautic.helper.email.address service to use in Replies constructor

escopecz commented on app/bundles/EmailBundle/Entity/StatRepository.php:
```diff
@@ -658,19 +658,20 @@ public function checkContactsSentEmail($contacts, $emailId)
      * @return array Formatted as [contactId => sentCount]
      */
-    public function getSentCountForContacts(array $contacts, $emailId)
+    public function getSentCountForContacts(array $contacts, array $emailIds)
```
This is a backward compatibility change. If a plugin use this method then it will cause an error when the second param was changed to an array from an int. I'd normally ask you to create a new method and mark this one as deprecated as I can see it's used only in one place. But since this will be releases in a major version we can introduce this BC change. However, it must be documented in https://github.com/mautic/mautic/blob/5.x/UPGRADE-5.0.md. Please do so.

escopecz reviewed (changes requested):
Please fix the issues reported by PHPSTAN and write tests to cover these changes. Let me know if you need some help.

kuzmany committed:
Fix ReplyTest

kuzmany committed:
Fix PHPStan

escopecz committed:
Converting PHP templates for the IntegrationsBundle to Twig

escopecz committed:
Removing files that are not needed anymore...

... those files were used when the bundle was an independent plugin

escopecz committed:
First batch of SMS twig refactoring

escopecz committed:
SMS uses also this page template, so added a Twig version of it too

escopecz committed:
More core templates have twig versions

escopecz committed:
Form template for SMS bundle refactored to Twig

escopecz committed:
json_decode Twig filter added

escopecz committed:
Rest of the SMS bundle PHP files refactored to Twig

escopecz committed:
Fixing form templates in the CoreBundle

escopecz committed:
Starting to use Twig templates instead of PHP templates

escopecz committed:
Fixing list of contacts on SMS detail page

escopecz committed:
Fixing SMS list

escopecz committed:
Fixing SMS timeline event template when error occurs

escopecz committed:
Using already existing "instanceof test" extension

escopecz committed:
Simplifying a couple of Twig extensions and starting integration tests for them

escopecz committed:
Adding more tests for the Twig extensions I touched in this PR

escopecz committed:
Fixing issues with sms type selection

kuzmany committed:
Fix date/datetime empty mysql expression for contact field condition

escopecz committed:
Fixing failing API-Library test for editing devices by PUT with ID that does not work

TypeError: method_exists(): Argument #1 ($object_or_class) must be of type object|string, null given

biozshock committed:
Updated Github actions to latest versions.

biozshock committed:
Updated Github set-output to setOutput via ENV var.

patrykgruszka committed:
Fix hourly email stats chart

npracht committed:
add more info in tooltip

annamunk committed:
GrapesJS rte disable fix for Mautic 5

annamunk committed:
GrapesJS build update

dennisameling committed:
Auto-update GrapesJS generated JS dist files

kuzmany committed:
Add $fieldType param in compareValue method as null by defaul

RCheesley commented:
@bergwerkweb please can you address the feedback in the PR so we can consider it for merging? Many thanks!

ynreob committed:
bugfix issue 11586

bergwerkweb committed:
Merge branch '5.x' into issue11586

escopecz commented:
There are many commits that do not belong here. I also fight with rebases/merges. Sometimes the simplest way is to create a fresh branch from the 5.x branch and cherry-pick the commits you need there.

bergwerkweb commented:
ok, i'll start from scratch....
